Allow using short type names for asset processors.

andriyDev opened the pull request:
# Objective

- Previously, asset processors needed to be **fully specified**. This meant writing met files by hand was incredibly cumbersome.
- In addition, using `core::any::type_name` is generally undesirable since it is only intended for debugging.

## Solution

- Require `TypePath` impl on `AssetLoader`, `AssetTransformer`, `AssetSaver`, and `AssetProcessor`.
- Register loaders and processors by their `TypePath::type_path`. This makes the path format stable.
- Additionally register the processors by their `TypePath::short_type_path`, and ensure there are no other processors with the same `short_type_path`. If a user tries to use an ambiguous short type path, we list out all full type paths.

Note: I left doing this with asset loaders as a future PR. There's more complexity there (since we have to deal with pre-registering asset loaders), and this seems like the bigger "bang for our buck".

## Testing

- I ran the asset_processing example. No change.
- I changed the asset processing example to use the short type path. It still behaved as expected.
- I created a new AssetTransformer in `asset_processing::sneaky::CoolTextTransformer` and a new processor that looked identical to the regular one (but using the sneaky transformer). It printed an error, listing out the full paths of both processors!

JMS55 reviewed (changes requested):
Code lgtm!

Please add a migration guide for the TypePath changes, and a release note entry showcasing this work.

Also, while I won't block on it, it would be great to have a unit test for ambiguous processor names. I'm worried that that is easy to break in future code changes.

andriyDev commented:
Fixed up CI, added the migration guide + release notes, and also added tests for getting the processor by name. We definitely need more testing around asset processing itself (I'm only testing that getting the processor by name all works), but that should be a different PR.

ChristopherBiscardi commented:
I think this is the first Bevy feature that will allow previously written assets (I guess .meta files in this case) to break when installing a new plugin. (ex: if that plugin also includes an ImageLoader, or CoolTextLoader in the example here). So the general guidance will probably become "namespace your loaders" like `MyCrateImageLoader` since this is now a global namespace, which doesn't really change the amount of handwritten code. We're moving the namespacing from the module path to a human-curated prefix.

This may also be introducing a difference in behavior between other types and loaders, ex: the [scenes example asset](https://github.com/bevyengine/bevy/blob/cd44b304cddf933631eea5e70acd348f49ddbd16/assets/scenes/load_scene_example.scn.ron#L10) currently requires full type paths for components, Resources, etc

I don't have a strong feeling either way (slight preference for adding new plugins not being able to break current applications), but I do see "why can I wrote loaders with shortnames but not components" coming up and we should have an answer for it.

hukasu commented:
Deriving `TypePath` should already namespace the type, no?

andriyDev commented:
> Deriving `TypePath` should already namespace the type, no?

Only if you use long type names. If you use short type names, then there's a risk that something like `ImageLoader` may be aliased by some other plugin that adds its own `ImageLoader`. So if you write your meta file to point to `ImageLoader`, it's more at risk of failing with an error, but if you name your loader `MyExtraSpecialImageLoader`, there's a smaller risk.

I still think this is a win. I don't think it's desirable for users to have to specify the module path since that can include info that doesn't matter to them, like `LoadTransformAndSave` is stored in `bevy_asset::processor`.

andriyDev commented:
The comment was that rather than calling their loaders `ImageLoader` the recommendation will be to "fake namespace" by saying `MyExtraSpecialImageLoader`. Which means the long name might look like: `my_extra_special_crate::MyExtraSpecialImageLoader`.

JMS55 commented:
Sorry, not going to have time to review this for a bit. Hopefully other community members can help with reviews.

andriyDev commented:
Rebased and added a test in lib.rs showing that short type names work!

ChristopherBiscardi reviewed (approved):
I am not personally convinced that introducing ambiguity here is a good idea, but the implementation seems good, there are lovely tests, and the functionality seems scoped to just loaders.

---

This PR is, in effect, creating a global namespace for loaders where Bevy has first priority to that namespace; Previously, everyone shared the crates.io/crate-names namespace. We could also make the loaders less arduous to write by setting the type paths to more human authorable lengths, such as `bevy_asset::LoadTransformAndSave`, which results in a still longer, but less prone to potential conflicts output.

```ron
(
    meta_format_version: "1.0",
    asset: Process(
        processor: "bevy_asset::LoadTransformAndSave<asset_processing::CoolTextLoader, asset_processing::CoolTextTransformer, asset_processing::CoolTextSaver>",
        settings: (
            loader_settings: (),
            transformer_settings: (),
            saver_settings: (),
        ),
    ),
)
```

All in all, I think the ecosystem will work around this when it comes up by namespacing loader names because everyone will want the short names but maybe that won't happen often enough to matter; so we take the short wins now at the risk of crates ending up with `CoolCrateImageLoader` later (instead of `cool_crate::ImageLoader`).

alice-i-cecile commented:
I'm similarly not fully convinced that this is the optimal design, but this is really well-made!

cart commented on crates/bevy_asset/src/processor/process.rs:
```diff
@@ -251,7 +259,7 @@ impl<P: Process> ErasedProcessor for P {
     fn default_meta(&self) -> Box<dyn AssetMetaDyn> {
         Box::new(AssetMeta::<(), P>::new(AssetAction::Process {
-            processor: core::any::type_name::<P>().to_string(),
+            processor: P::type_path().to_string(),
```
I think we probably want to switch this to the short name so the auto-generated metadata uses the more readable names.

cart replied:
Given that meta files are intended to be auto-generated (at least initially), and not manually defined by developers, what that would mean in practice is that short names are never used. I personally believe short paths should be the default behavior, as they improve legibility significantly, and also make it easier to for developers to step in and make manual edits. Otherwise I don't see the point in introducing this complexity and new error path, as it won't move the needle in the 99.99% case.

That being said, this should take into account the current set of loaders to determine whether or not to use the full type path to disambiguate. Otherwise (as you said), in the context of conflicting paths, this would generate unloadable data.

In the editor, we'll also want to be able to, for a given meta file, resolve ambiguities (perhaps via a UI prompt) and re-save. This is because meta files could be defined in other projects or contexts that are not ambiguous.

Given that current limitation, it does make sense to go back to the full path in the short term. But I'd like us to be aligned on the intended scope of this feature before merging.

cart reviewed (approved):
This is the right path forward. `TypePath` was implemented with this scenario in mind, and this approach is the way I imagined it would be used.
